chromadb/api/models/AsyncCollection.py

```diff
@@ -0,0 +1,470 @@
+from typing import (
+    TYPE_CHECKING,
+    Optional,
+    Union,
+)
+import numpy as np
+
+from chromadb.api.types import (
+    URI,
+    CollectionMetadata,
+    Embedding,
+    Include,
+    Metadata,
+    Document,
+    Image,
+    Where,
+    IDs,
+    GetResult,
+    QueryResult,
+    ID,
+    OneOrMany,
+    WhereDocument,
+    maybe_cast_one_to_many_ids,
+    maybe_cast_one_to_many_embedding,
+    maybe_cast_one_to_many_document,
+    maybe_cast_one_to_many_image,
+    maybe_cast_one_to_many_uri,
+    validate_ids,
+    validate_include,
+    validate_metadata,
+    validate_where,
+    validate_where_document,
+    validate_n_results,
+    validate_embeddings,
+)
+
+from .CollectionCommon import CollectionCommon
+
+if TYPE_CHECKING:
+    from chromadb.api import AsyncServerAPI  # noqa: F401
+
+
+class AsyncCollection(CollectionCommon["AsyncServerAPI"]):
+    async def add(
+        self,
+        ids: OneOrMany[ID],
+        embeddings: Optional[
+            Union[
+                OneOrMany[Embedding],
+                OneOrMany[np.ndarray],
+            ]
+        ] = None,
+        metadatas: Optional[OneOrMany[Metadata]] = None,
+        documents: Optional[OneOrMany[Document]] = None,
+        images: Optional[OneOrMany[Image]] = None,
+        uris: Optional[OneOrMany[URI]] = None,
+    ) -> None:
+        """Add embeddings to the data store.
+        Args:
+            ids: The ids of the embeddings you wish to add
+            embeddings: The embeddings to add. If None, embeddings will be computed based on the documents or images using the embedding_function set for the Collection. Optional.
+            metadatas: The metadata to associate with the embeddings. When querying, you can filter on this metadata. Optional.
+            documents: The documents to associate with the embeddings. Optional.
+            images: The images to associate with the embeddings. Optional.
+            uris: The uris of the images to associate with the embeddings. Optional.
+
+        Returns:
+            None
+
+        Raises:
+            ValueError: If you don't provide either embeddings or documents
+            ValueError: If the length of ids, embeddings, metadatas, or documents don't match
+            ValueError: If you don't provide an embedding function and don't provide embeddings
+            ValueError: If you provide both embeddings and documents
+            ValueError: If you provide an id that already exists
+
+        """
+        (
+            ids,
+            embeddings,
+            metadatas,
+            documents,
+            images,
+            uris,
+        ) = self._validate_embedding_set(
+            ids, embeddings, metadatas, documents, images, uris
+        )
+
+        # We need to compute the embeddings if they're not provided
+        if embeddings is None:
```

Comment:
I would prefer if we found a way to code share the logic in these methods

Reply:
[d3a059b](https://github.com/chroma-core/chroma/pull/2297/commits/d3a059be5be4ae029bc001eac3e2694653e35e95)